README
NAME
    Plack::Middleware::Image::Scale - Resize jpeg and png images on the fly

VERSION
    version 0.009

SYNOPSIS
        ## example1.psgi
    
        builder {
            enable 'ConditionalGET';
            enable 'Image::Scale';
            enable 'Static', path => qr{^/images/};
            $app;
        };

    A request to /images/foo_40x40.png will use images/foo.(png|jpg|gif) as
    original, scale it to 40x40 px size and convert to PNG format.

        ## example2.psgi

        my $thumber = builder {
            enable 'ConditionalGET';
            enable 'Image::Scale',
                width => 200, height => 100,
                flags => { fill => 'ff00ff' };
            Plack::App::File->new( root => 'images' );
        };

        builder {
            mount '/thumbs' => $thumber;
            mount '/' => $app;
        };

    A request to /thumbs/foo.png will use images/foo.(png|jpg|gif) as
    original, scale it small enough to fit 200x100 px size, fill extra
    borders (top/down or left/right, depending on the original image aspect
    ratio) with cyan background, and convert to PNG format. Also clipping is
    available, see "CONFIGURATION".

        ## see example4.psgi

        my %imagesize = Config::General->new('imagesize.conf')->getall;
    
        # ...
    
        enable 'Image::Scale', size => \%imagesize;

    A request to /images/foo_medium.png will use images/foo.(png|jpg|gif) as
    original. The size and flags are taken from the configuration file as
    parsed by Config::General.

        ## imagesize.conf

        <medium>
            width   200
            height  100
            crop
        </medium>
        <big>
            width   300
            height  100
            crop
        </big>
        <thumbred>
            width   50
            height  100
            fill    ff0000
        </thumbred>

    For more examples, browse into directory eg
    <http://cpansearch.perl.org/src/PNU/> inside the distribution directory
    for this version.

DESCRIPTION
    Scale and convert images to the requested format on the fly. By default
    the size and other scaling parameters are extracted from the request
    URI. Scaling is done with Image::Scale.

    The original image is not modified or even accessed directly by this
    module. The converted image is not cached, but the request can be
    validated (If-Modified-Since) against original image without doing the
    image processing. This middleware should be used together a cache proxy,
    that caches the converted images for all clients, and implements content
    validation.

    The response headers (like Last-Modified or ETag) are from the original
    image, but body is replaced with a PSGI content filter to do the image
    processing. The original image is fetched from next middleware layer or
    application with a normal PSGI request. You can use
    Plack::Middleware::Static, or Catalyst::Plugin::Static::Simple for
    example.

    See "CONFIGURATION" for various size/format specifications that can be
    used in the request URI, and "ATTRIBUTES" for common configuration
    options that you can use when constructing the middleware.

ATTRIBUTES
  path
    Must be a RegexpRef, CodeRef, Str or Undef.

    The PATH_INFO is compared against this value to evaluate if the request
    should be processed. Undef (the default) will match always. "PATH_INFO"
    is topicalized by settings it to $_, and it may be rewritten during
    "CodeRef" matching. Rewriting can be used to relocate image paths, much
    like "path" parameter for Plack::Middleware::Static.

    If path matches, next it will be compared against "name". If path
    doesn't match, the request will be delegated to the next middleware
    layer or application.

  match
    Must be a RegexpRef, or CodeRef.

    The PATH_INFO, possibly rewritten during "path" matching, is compared
    against this value to extract "name", "size" and "ext". The default
    value is:

        qr{^(.+)(?:_(.+?))?(?:\.(jpe?g|png|image))$}

    The expression is evaluated in array context and may return three
    elements: "name", "size" and "ext". Returning an empty array means no
    match. Non-matching requests are delegated to the next middleware layer
    or application.

    If the path matches, the original image is fetched from
    "name"."orig_ext", scaled with parameters extracted from "size" and
    converted to the content type defined by "ext". See also "any_ext".

  size
    Must be a RegexpRef, CodeRef, HashRef, Undef.

    The "size" extracted by "match" is compared against this value to
    evaluate if the request should be processed, and to map it into width,
    height and flags for image processing. Undef will match always and use
    default width, height and flags as defined by the "ATTRIBUTES". The
    default value is:

        qr{^(\d+)?x(\d+)?(?:-(.+))?$}

    The expression is evaluated in array context and may return three
    elements; "width", "height" and "flags". Returning an empty array means
    no match. Non-matching requests are delegated to the next middleware
    layer or application.

    Optionally a hash reference can be returned. Keys "width", "height", and
    any remaining keys as an hash reference, will be unrolled from the hash
    reference.

  any_ext
    If defined and request "ext" is equal to this, the content type of the
    original image is used in the output. This means that the image format
    of the original image is preserved. Default is "image".

  orig_ext
    ArrayRef of possible original image formats. See "fetch_orig".

  memory_limit
    Memory limit for the image scaling in bytes, as defined in Image::Scale.

  jpeg_quality
    JPEG quality, as defined in Image::Scale.

  width
    Use this to set and override image width.

  height
    Use this to set and override image height.

  flags
    Use this to set and override image processing flags.

METHODS
  fetch_orig
    Call parameters: PSGI request HashRef $env, Str $basename. Return value:
    PSGI response ArrayRef $res.

    The original image is fetched from the next layer or application. All
    possible extensions defined in "orig_ext" are tried in order, to search
    for the original image. All other responses except a straight 404 (as
    returned by Plack::Middleware::Static for example) are considered
    matches.

  body_scaler
    Call parameters: @args. Return value: PSGI content filter CodeRef $cb.

    Create the content filter callback and return a CodeRef to it. The
    filter will buffer the data and call "image_scale" with parameters @args
    when EOF is received, and finally return the converted data.

  image_scale
    Call parameters: ScalarRef $buffer, String $ct, Int $width, Int $height,
    HashRef|Str $flags. Return value: $imagedata

    Read image from $buffer, scale it to $width x $height and return as
    content-type $ct. Optional $flags to specify image processing options
    like background fills or cropping.

CONFIGURATION
    The default match pattern for URI is
    "*...*_*width*x*height*-*flags*.*ext*".

    If URI doesn't match, the request is passed through. Any number of flags
    can be specified, separated with "-". Flags can be boolean (exists or
    doesn't exist), or have a numerical value. Flag name and value are
    separated with a zero-width word to number boundary. For example "z20"
    specifies flag "z" with value 20.

  width
    Width of the output image. If not defined, it can be anything (to
    preserve the image aspect ratio).

  height
    Height of the output image. If not defined, it can be anything (to
    preserve the image aspect ratio).

  flags: fill
    Image aspect ratio is preserved by scaling the image to fit within the
    specified size. This means scaling to the smaller or the two possible
    sizes that preserve aspect ratio. Extra borders of background color are
    added to fill the requested image size exactly.

        /images/foo_400x200-fill.png

    If fill has a value, it specifies the background color to use. Undefined
    color with png output means transparent background.

  flags: crop
    Image aspect ratio is preserved by scaling and cropping from middle of
    the image. This means scaling to the bigger of the two possible sizes
    that preserve the aspect ratio, and then cropping to the exact size.

  flags: fit
    Image aspect ratio is preserved by scaling the image to the smaller of
    the two possible sizes. This means that the resulting picture may have
    one dimension smaller than specified, but cropping or filling is
    avoided.

    See documentation in distribution directory "doc" for a visual
    explanation.

  flags: z
    Zoom the original image N percent bigger. For example "z20" to zoom 20%.
    Zooming applies only to explicitly defined width and/or height, and it
    does not change the crop size.

        /images/foo_40x-z20.png

CAVEATS
    The cropping requires Imager. This is a run-time dependency, and
    fallback is not to crop the image to the expected size.

SEE ALSO
    Image::Scale

    Imager

    Plack::App::ImageMagick

AUTHOR
    Panu Ervamaa <pnu@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Panu Ervamaa.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changelog for Plack::Middleware::Image::Scale

{{$NEXT}}

0.011     2016-02-04 12:50:50+02:00 Europe/Helsinki

    * Fixes RT#90946 Class::MOP::load_class deprecation (Zoffix Znet)
    * Cleanup whitespaces and various warnings
    * Update build system, rename README -> README.md, fix synopsis

0.010     2012-11-07 12:26:00 Europe/Helsinki

    * Also match jpeg file extension.

0.009     2012-08-29 14:03:12 Europe/Helsinki

    * Experimental and optional PDF support, requires
      Image::Magick::Thumbnail::PDF.

0.008     2011-08-24 13:46:16 Europe/Helsinki

    * Implement new flag "fit", additional documentation in doc directory.
    * Do not try to process empty bodies.

0.007     2011-04-11 21:09:42 Europe/Helsinki

    * Changes to matching logic. See README for overview.
    * Cleanup documentation.
    * Learn any_ext attribute (default "image").
    * Use $self->response_cb instead of Plack::Util::response_cb.
    * Better test coverage.

0.006     2011-04-05 19:02:18 Europe/Helsinki

    * Size matching can read HashRef of HashRefs for easy
      Config::General style configuration data support.

0.005     2011-04-04 17:37:41 Europe/Helsinki

    * Add examples for using a CodeRef for path parameter.

0.004     2011-04-04 00:51:31 Europe/Helsinki

    * Override width, height and flags with optional module attributes.
    * Easier to extend: "match_*" replaced with "path" Regexp/CodeRef.

0.003     2011-04-01 21:59:11 Europe/Helsinki

    * Documentation changes. First non-trial release.

0.002     2011-03-31 09:52:21 Europe/Helsinki

    * Documentation changes.

0.001     2011-03-30 21:29:17 Europe/Helsinki

    * Initial release.



t/lib/Test/Invocation/Arguments.pm
package Test::Invocation::Arguments;
# ABSTRACT: Capture method invocation arguments
use Moose;
use Class::MOP;
use Data::Dumper;

has class  => ( is => 'ro', isa => 'ClassName' );
has method => ( is => 'ro', isa => 'Str' );
has method_orig => ( is => 'rw' );

has _log => (
    isa     => 'ArrayRef[ArrayRef]',
    default => sub { [] },
    traits  => ['Array'],
    handles => {
        push   => 'push',
        pop    => 'pop',
        count  => 'count',
    }
);

sub BUILD {
    my $self = shift;
    my $meta = Class::MOP::Class->initialize( $self->class );
    $self->method_orig( $meta->get_method( $self->method ) );
    $meta->add_before_method_modifier( $self->method, sub {
        my $invocant = shift;
        $self->push( \@_ );
        ## TODO should make a deep copy first
    } );
}

sub DEMOLISH {
    my $self = shift;
    my $meta = Class::MOP::Class->initialize( $self->class );
    $meta->add_method( $self->method, $self->method_orig );
}

1;


lib/Plack/Middleware/Image/Scale.pm
use strict;
package Plack::Middleware::Image::Scale;
# ABSTRACT: Resize jpeg and png images on the fly

use Moose;
use Class::Load;
use Plack::Util;
use Plack::MIME;
use Try::Tiny;
use Image::Scale;
use List::Util qw( min max );
use Carp;

extends 'Plack::Middleware';

# VERSION

=attr path

Must be a L<RegexpRef|Moose::Util::TypeConstraints/Default_Type_Constraints>,
L<CodeRef|Moose::Util::TypeConstraints/Default_Type_Constraints>,
L<Str|Moose::Util::TypeConstraints/Default_Type_Constraints> or
L<Undef|Moose::Util::TypeConstraints/Default_Type_Constraints>.

The L<PATH_INFO|PSGI/The_Environment> is compared against this value to
evaluate if the request should be processed. Undef (the default) will match
always.  C<PATH_INFO> is topicalized by settings it to C<$_>, and it may be
rewritten during C<CodeRef> matching. Rewriting can be used to relocate image
paths, much like C<path> parameter for L<Plack::Middleware::Static>.

If path matches, next it will be compared against L</name>. If path doesn't
match, the request will be delegated to the next middleware layer or
application.

=cut

has path => (
    is => 'rw', lazy => 1, isa => 'RegexpRef|CodeRef|Str|Undef',
    default => undef
);

=attr match

Must be a L<RegexpRef|Moose::Util::TypeConstraints/Default_Type_Constraints>,
or L<CodeRef|Moose::Util::TypeConstraints/Default_Type_Constraints>.

The L<PATH_INFO|PSGI/The_Environment>, possibly rewritten during L</path>
matching, is compared against this value to extract C<name>, C<size>
and C<ext>. The default value is:

    qr{^(.+)(?:_(.+?))?(?:\.(jpe?g|png|image))$}

The expression is evaluated in array context and may return three elements:
C<name>, C<size> and C<ext>. Returning an empty array means no match.
Non-matching requests are delegated to the next middleware layer or
application.

If the path matches, the original image is fetched from C<name>.L</orig_ext>,
scaled with parameters extracted from C<size> and converted to the content type
defined by C<ext>. See also L</any_ext>.

=cut

has match => (
    is => 'rw', lazy => 1, isa => 'RegexpRef|CodeRef',
    default => sub { qr{^(.+?)(?:_([^_]+?))?(?:\.(jpe?g|png|image))$} }
);

=attr size

Must be a L<RegexpRef|Moose::Util::TypeConstraints/Default_Type_Constraints>,
L<CodeRef|Moose::Util::TypeConstraints/Default_Type_Constraints>,
L<HashRef|Moose::Util::TypeConstraints/Default_Type_Constraints>,
L<Undef|Moose::Util::TypeConstraints/Default_Type_Constraints>.

The C<size> extracted by L</match> is compared against this value to evaluate
if the request should be processed, and to map it into width, height and flags
for image processing. Undef will match always and use default width, height
and flags as defined by the L</ATTRIBUTES>. The default value is:

    qr{^(\d+)?x(\d+)?(?:-(.+))?$}

The expression is evaluated in array context and may return three elements;
C<width>, C<height> and C<flags>. Returning an empty array means no match.
Non-matching requests are delegated to the next middleware layer or
application.

Optionally a hash reference can be returned. Keys C<width>, C<height>, and any
remaining keys as an hash reference, will be unrolled from the hash reference.

=cut

has size => (
    is => 'rw', lazy => 1, isa => 'RegexpRef|CodeRef|HashRef|Undef',
    default => sub { qr{^(\d+)?x(\d+)?(?:-(.+))?$} }
);

=attr any_ext

If defined and request C<ext> is equal to this, the content type of the original
image is used in the output. This means that the image format of the original
image is preserved. Default is C<image>.

=cut

has any_ext => (
    is => 'rw', lazy => 1, isa => 'Str|Undef',
    default => 'image'
);

=attr orig_ext

L<ArrayRef|Moose::Util::TypeConstraints/Default_Type_Constraints>
of possible original image formats. See L</fetch_orig>.

=cut

has orig_ext => (
    is => 'rw', lazy => 1, isa => 'ArrayRef',
    default => sub { [qw( jpg png gif jpeg )] }
);

=attr memory_limit

Memory limit for the image scaling in bytes, as defined in
L<Image::Scale|Image::Scale/resize(_\%OPTIONS_)>.

=cut

has memory_limit => (
    is => 'rw', lazy => 1, isa => 'Int|Undef',
    default => 10_000_000 # bytes
);

=attr jpeg_quality

JPEG quality, as defined in
L<Image::Scale|Image::Scale/as_jpeg(_[_$QUALITY_]_)>.

=cut

has jpeg_quality => (
    is => 'rw', lazy => 1, isa => 'Int|Undef',
    default => undef
);

=attr width

Use this to set and override image width.

=cut

has width => (
    is => 'rw', lazy => 1, isa => 'Int|Undef',
    default => undef
);

=attr height

Use this to set and override image height.

=cut

has height => (
    is => 'rw', lazy => 1, isa => 'Int|Undef',
    default => undef
);

=attr flags

Use this to set and override image processing flags.

=cut

has flags => (
    is => 'rw', lazy => 1, isa => 'HashRef|Undef',
    default => undef
);

sub call {
    my ($self,$env) = @_;
    my $path = $env->{PATH_INFO};
    my @param;

    if ( defined $self->path ) {
        my ($m) = _match($path,$self->path);
        return $self->app->($env) unless $m;
    }

    my @m = _match($path,$self->match);
    return $self->app->($env) unless @m;
    ($path, my $size, my $ext) = @m;
    return $self->app->($env) unless $path and $ext;

    if ( defined $size ) {
        @param = _unroll(_match($size,$self->size));
        return $self->app->($env) unless @param;
    }

    my $res = $self->fetch_orig($env,$path);
    return $self->app->($env) unless $res;

    ## Post-process the response with a body filter
    $self->response_cb( $res, sub {
        my $res = shift;
        my $orig_ct = Plack::Util::header_get( $res->[1], 'Content-Type' );
        my $ct;
        if ( defined $self->any_ext and $ext eq $self->any_ext ) {
            $ct = Plack::Util::header_get( $res->[1], 'Content-Type' );
        } else {
            $ct = Plack::MIME->mime_type(".$ext");
            Plack::Util::header_set( $res->[1], 'Content-Type', $ct );
        }
        return $self->body_scaler( $ct, $orig_ct, @param );
    });
}

## Helper for matching a Scalar value against CodeRef, HashRef,
## RegexpRef or Str. The first argument may be modified during match.
sub _match {
    my @match;
    for ( $_[0] ) {
        my $match = $_[1];
        @match =
          'CODE' eq ref $match ? $match->($_) :
          'HASH' eq ref $match ? $match->{$_} :
        'Regexp' eq ref $match ? $_ =~ $match :
                defined $match ? (substr($_,0,length $match) eq $match ? ($match) : ()) :
                                 undef;
    }
    return @match;
}

## Helper for extracting (width,height,flags) from
## HashRef or ArrayRef.
sub _unroll {
    return unless @_;
    for ( $_[0] ) {
        ## Config::General style hash of hashrefs.
        if ( ref eq 'HASH' ) {
            my %e = %{$_};
            return (delete @e{'width','height'}, \%e);
        ## Manual config friendly hash of arraysrefs.
        } elsif ( ref eq 'ARRAY' ) {
            return @{$_};
        }
    }
    return @_;
}

=method fetch_orig

Call parameters: PSGI request HashRef $env, Str $basename.
Return value: PSGI response ArrayRef $res.

The original image is fetched from the next layer or application.  All
possible extensions defined in L</orig_ext> are tried in order, to search for
the original image. All other responses except a straight 404 (as returned by
L<Plack::Middleware::Static> for example) are considered matches.

=cut

sub fetch_orig {
    my ($self,$env,$basename) = @_;

    for my $ext ( @{$self->orig_ext} ) {
        local $env->{PATH_INFO} = "$basename.$ext";
        my $r = $self->app->($env);
        return $r unless ref $r eq 'ARRAY' and $r->[0] == 404;
    }
    return;
}

=method body_scaler

Call parameters: @args. Return value: PSGI content filter CodeRef $cb.

Create the content filter callback and return a CodeRef to it. The filter will
buffer the data and call L</image_scale> with parameters C<@args> when EOF is
received, and finally return the converted data.

=cut

sub body_scaler {
    my $self = shift;
    my @args = @_;

    my $buffer = q{};
    my $filter_cb = sub {
        my $chunk = shift;

        ## Buffer until we get EOF
        if ( defined $chunk ) {
            $buffer .= $chunk;
            return q{}; #empty
        }

        ## Return EOF when done
        return if not defined $buffer;

        ## Process the buffer
        my $img = $buffer ? $self->image_scale(\$buffer,@args) : '';
        undef $buffer;
        return $img;
    };

    return $filter_cb;
}

=method image_scale

Call parameters: ScalarRef $buffer, String $ct, Int $width, Int $height, HashRef|Str $flags.
Return value: $imagedata

Read image from $buffer, scale it to $width x $height and
return as content-type $ct. Optional $flags to specify image processing
options like background fills or cropping.

=cut

sub image_scale {
    my ($self, $bufref, $ct, $orig_ct, $width, $height, $flags) = @_;

    ## $flags can be a HashRef, or it's parsed as a string
    my %flag = 'HASH' eq ref $flags ? %{ $flags } :
    map { (split /(?<=\w)(?=\d)/, $_, 2)[0,1]; } split '-', $flags || '';

    $width  = $self->width      if defined $self->width;
    $height = $self->height     if defined $self->height;
    %flag   = %{ $self->flags } if defined $self->flags;

    my $owidth  = $width;
    my $oheight = $height;

    if ( defined $flag{z} and $flag{z} > 0 ) {
        $width  *= 1 + $flag{z} / 100 if $width;
        $height *= 1 + $flag{z} / 100 if $height;
    }

    my $output;
    if (defined $orig_ct and $orig_ct eq 'application/pdf') {
        try {
            Class::Load::load_class('Image::Magick::Thumbnail::PDF');
            Class::Load::load_class('File::Temp');
            my $in = File::Temp->new( SUFFIX => '.pdf' );
            my $out = File::Temp->new( SUFFIX => '.png' );
            $in->write( $$bufref ); $in->close;
            Image::Magick::Thumbnail::PDF::create_thumbnail(
                $in->filename, $out->filename, $flag{p}||1, {
                    frame => 0, normalize => 0,
                    restriction => max($width, $height),
                }
            );
            my $pdfdata;
            $out->seek( 0, 0 );
            $out->read( $pdfdata, 9999999 );
            $bufref = \$pdfdata;
        } catch {
            carp $_;
            $output = $$bufref;
        };
    }
    try {
        my $img = Image::Scale->new($bufref)
            or die 'Invalid data / image format not recognized';

        if ( exists $flag{crop} and defined $width and defined $height ) {
            my $ratio = $img->width / $img->height;
            $width  = max $width , $height * $ratio;
            $height = max $height, $width / $ratio;
        } elsif ( exists $flag{fit} and defined $width and defined $height ) {
            my $ratio = $img->width / $img->height;
            $width  = min $width , $height * $ratio;
            $height = min $height, $width / $ratio;
        }

        unless ( defined $width or defined $height ) {
            ## We want to keep the size, but Image::Scale
            ## doesn't return data unless we call resize.
            $width = $img->width; $height = $img->height;
        }
        $img->resize({
            defined $width  ? (width  => $width)  : (),
            defined $height ? (height => $height) : (),
            exists  $flag{fill} ? (keep_aspect => 1) : (),
            defined $flag{fill} ? (bgcolor => hex $flag{fill}) : (),
            defined $self->memory_limit ?
                (memory_limit => $self->memory_limit) : (),
        });

        $output = $ct eq 'image/jpeg' ? $img->as_jpeg($self->jpeg_quality || ()) :
                  $ct eq 'image/png'  ? $img->as_png :
                  die "Conversion to '$ct' is not implemented";
    } catch {
        carp $_;
        $output = $$bufref;
    };

    if ( defined $owidth  and $width  > $owidth or
         defined $oheight and $height > $oheight ) {
        try {
            Class::Load::load_class('Imager');
            my $img = Imager->new;
            $img->read( data => $output ) || die;
            my $crop = $img->crop(
                defined $owidth  ? (width  => $owidth)  : (),
                defined $oheight ? (height => $oheight) : (),
            );
            $crop->write( data => \$output, type => (split '/', $ct)[1] );
        } catch {
            carp $_;
        };
    }

    return $output;
}

1;

=head1 SYNOPSIS

    ## example1.psgi
    use Plack::Builder;
    use Plack::Middleware::Image::Scale;
    my $app = sub { return [200,[],[]] };

    builder {
        enable 'ConditionalGET';
        enable 'Image::Scale';
        enable 'Static', path => qr{^/images/};
        $app;
    };

A request to /images/foo_40x40.png will use images/foo.(png|jpg|gif|jpeg) as
original, scale it to 40x40 px size and convert to PNG format.

    ## example2.psgi
    use Plack::Builder;
    use Plack::App::File;
    use Plack::Middleware::Image::Scale;
    my $app = sub { return [200,['Content-Type'=>'text/plain'],['hello']] };

    my $thumber = builder {
        enable 'ConditionalGET';
        enable 'Image::Scale',
            width => 200, height => 100,
            flags => { fill => 'ff00ff' };
        Plack::App::File->new( root => 'images' );
    };

    builder {
        mount '/thumbs' => $thumber;
        mount '/' => $app;
    };

A request to /thumbs/foo.png will use images/foo.(png|jpg|gif|jpeg) as original,
scale it small enough to fit 200x100 px size, fill extra borders (top/down or
left/right, depending on the original image aspect ratio) with cyan
background, and convert to PNG format. Also clipping is available, see
L</CONFIGURATION>.

=head1 DESCRIPTION

Scale and convert images to the requested format on the fly. By default the
size and other scaling parameters are extracted from the request URI.  Scaling
is done with L<Image::Scale>.

The original image is not modified or even accessed directly by this module.
The converted image is not cached, but the request can be validated
(If-Modified-Since) against original image without doing the image processing.
This middleware should be used together a cache proxy, that caches the
converted images for all clients, and implements content validation.

The response headers (like Last-Modified or ETag) are from the original image,
but body is replaced with a PSGI L<content
filter|Plack::Middleware/RESPONSE_CALLBACK> to do the image processing.  The
original image is fetched from next middleware layer or application with a
normal PSGI request. You can use L<Plack::Middleware::Static>, or
L<Catalyst::Plugin::Static::Simple> for example.

See L</CONFIGURATION> for various size/format specifications that can be used
in the request URI, and L</ATTRIBUTES> for common configuration options
that you can use when constructing the middleware.

=head1 CONFIGURATION

The default match pattern for URI is
"I<...>_I<width>xI<height>-I<flags>.I<ext>".

If URI doesn't match, the request is passed through. Any number of flags can
be specified, separated with C<->.  Flags can be boolean (exists or doesn't
exist), or have a numerical value. Flag name and value are separated with a
zero-width word to number boundary. For example C<z20> specifies flag C<z>
with value C<20>.

=head2 width

Width of the output image. If not defined, it can be anything
(to preserve the image aspect ratio).

=head2 height

Height of the output image. If not defined, it can be anything
(to preserve the image aspect ratio).

=head2 flags: fill

Image aspect ratio is preserved by scaling the image to fit within the
specified size. This means scaling to the smaller or the two possible sizes
that preserve aspect ratio.  Extra borders of background color are added to
fill the requested image size exactly.

    /images/foo_400x200-fill.png

If fill has a value, it specifies the background color to use. Undefined color
with png output means transparent background.

=head2 flags: crop

Image aspect ratio is preserved by scaling and cropping from middle of the
image. This means scaling to the bigger of the two possible sizes that
preserve the aspect ratio, and then cropping to the exact size.

=head2 flags: fit

Image aspect ratio is preserved by scaling the image to the smaller of the two
possible sizes. This means that the resulting picture may have one dimension
smaller than specified, but cropping or filling is avoided.

See documentation in distribution directory C<doc> for a visual explanation.

=head2 flags: z

Zoom the original image N percent bigger. For example C<z20> to zoom 20%.
Zooming applies only to explicitly defined width and/or height, and it does
not change the crop size.

    /images/foo_40x-z20.png

=head1 EXAMPLES

    ## see example4.psgi

    my %imagesize = Config::General->new('imagesize.conf')->getall;

    # ...

    enable 'Image::Scale', size => \%imagesize;

A request to /images/foo_medium.png will use images/foo.(png|jpg|gif|jpeg) as
original. The size and flags are taken from the configuration file as
parsed by Config::General.

    ## imagesize.conf

    <medium>
        width   200
        height  100
        crop
    </medium>
    <big>
        width   300
        height  100
        crop
    </big>
    <thumbred>
        width   50
        height  100
        fill    ff0000
    </thumbred>

For more examples, browse into directory
L<eg|http://cpansearch.perl.org/src/PNU/> inside the distribution
directory for this version.

=head1 CAVEATS

The cropping requires L<Imager>. This is a run-time dependency, and
fallback is not to crop the image to the expected size.

=head1 SEE ALSO

L<Image::Scale>

L<Imager>

L<Plack::App::ImageMagick>

=cut
